wandb/sdk/lib/redirect.py

```diff
@@ -347,28 +379,40 @@ def _handle_csi(self, csi, params, command):
 
     def _get_line(self, n, formatting=True):
         line = self.buffer[n]
+        line_len = self._get_line_len(n)
         if not formatting:
-            return "".join(
-                [self.buffer[line][j] for j in range(self._get_line_len(line))]
-            )
+            return "".join([line[j].data for j in range(line_len)])
         else:
-            out = ""
-            prev_char = _defchar
-            for i in range(self._get_line_len(n)):
-                c = line[i]
-                if c.fg != prev_char.fg:
-                    out += _get_char(c.fg)
-                if c.bg != prev_char.bg:
-                    out += _get_char(c.bg)
-                for k in c.__slots__[3:]:
-                    ck = c[k]
-                    if ck != prev_char[k]:
-                        if not ck:
-                            k = "/" + k
-                        out += _get_char(ANSI_STYLES_REV[k])
-                out += c.data
-                prev_char = c
-            return out
+            out = [line[i].data for i in range(line_len)]
```

Comment:
The code below is GNARLEY.  Let's add some more comments about what is happening here.